qcodes/instrument_drivers/Lakeshore/lakeshore_base.py

```diff
@@ -0,0 +1,302 @@
+        `range_limits` is the upper limit for using a certain heater current
+        """
+        if self.range_limits.get_latest() is None:
+            raise RuntimeError('Error when calling set_range_from_temperature:'
+                               ' You must specify the output range limits '
+                               'before automatically setting the range '
+                               '(e.g. inst.range_limits([0.021, 0.1, 0.2, '
+                               '1.1, 2, 4, 8]))')
+        i = bisect(self.range_limits.get_latest(), temperature)
+        # there is a `+1` because `self.RANGES` includes `'off'` as the first
+        # value.
+        self.output_range(self.INVERSE_RANGES[i+1])
+        return self.output_range()
+
+    def set_setpoint_and_range(self, temperature):
+        # TODO: Range should be selected according to current temperature,
+        # not according to current setpoint
+        self.set_range_from_temperature(temperature)
+        self.setpoint(temperature)
+
+    def wait_until_set_point_reached(self, *,wait_cycle_time: float=None,
+                                     wait_tolerance: float=None,
+                                     wait_equilibration_time: float=None):
+        wait_cycle_time =  wait_cycle_time or self.wait_cycle_time.get_latest()
+        wait_tolerance = wait_tolerance or self.wait_tolerance.get_latest()
+        wait_equilibration_time = (wait_equilibration_time or
+                                   self.wait_equilibration_time.get_latest())
+        active_channel_id = self.input_channel()
+        active_channel = self.root_instrument.channels[active_channel_id]
+        # assert active_channel._channel == active_channel_id
+
+        t_setpoint = self.setpoint()
+        t_reading = active_channel.temperature()
+        start_time_in_tolerance_zone = None
+        is_in_tolerance_zone = False
+        while True:
+            t_reading = active_channel.temperature()
+            log.debug(f'loop iteration with t reading of {t_reading}')
+            # if temperature is lower than sensor range, keep on waiting
+            # TODO(DV):only do this coming from one direction
+            if t_reading:
+                delta = abs(t_reading-t_setpoint)/t_reading
+                if delta < wait_tolerance:
+                    if is_in_tolerance_zone:
+                        if (time.monotonic() - start_time_in_tolerance_zone
+                            > wait_equilibration_time):
+                            break
+                    else:
+                        start_time_in_tolerance_zone = time.monotonic()
+                        is_in_tolerance_zone = True
+                else:
+                    if is_in_tolerance_zone:
+                        is_in_tolerance_zone = False
+                        start_time_in_tolerance_zone = None
+
+            log.debug(f'waiting to reach setpoint: temp at '
+                      f'{t_reading}, delta:{delta}')
```

Comment:
this line throws exception about unknown variable `delta` in case `t_reading` is `None`/`False`/etc. Also, why would we have `if t_reading:` condition in the first place?

Reply:
fixed, `if t_reading:` is gone